vstart.sh: run multiple rgws with different ids

jecluis opened the pull request:
We should expect being able to run multiple rgws with different ids
using vstart, and yet that was not the case as they were sharing the
same name (and same key).

We now mimic each rgw running on a different host, with ids thus being
'client.rgw.hostN', with N being A to Z. For simplicity's sake, we're
bounding the number of rgws to 26.

Signed-off-by: Joao Eduardo Luis \<joao@suse.de>

theanalyst commented:
jenkins test this please

jecluis commented:
jenkins test this please

jecluis commented:
I really can't see why this is failing the checks. The log has them all passing and then blowing up on jenkins ¯\\_(ツ)_/¯

jecluis commented:
@cbodley can you take a look at this? would be nice to get this into vstart imo.

cbodley commented on src/vstart.sh:
```diff
@@ -464,6 +471,40 @@ get_pci_selector() {
+        rgw crypt require ssl = false
+        ; uncomment the following to set LC days as the value in seconds;
+        ; needed for passing lc time based s3-tests (can be verbose)
+        ; rgw lc debug interval = 10
```
the `rgw frontends` and `admin socket` are different between instances, but the rest could stay under `[client.rgw]`

jecluis replied:
I believe the `admin socket` will be different for each instance (given we're specifying dependent on `${rgw_id}`, but yeah, we're assigning the same port to each instance, which will be bad. Would an incremental port be okay, or this have some other behavior?

cbodley replied:
incremental port is how it worked before, and it looks like you're incrementing current_port here too :+1: 

i'm just suggesting that we leave the initial [client.rgw] section for the variables that don't change between instances, and only put `rgw frontends` and `admin socket` in these [client.rgw.$id] sections

cbodley commented:
@jecluis what exactly are you looking to do, that needs the radosgws to run with separate ids? vstart.sh is overriding the `--admin-socket` and `--rgw_frontends` on the radosgw command line, so we don't rely on ceph.conf for those at all

there was some good discussion long ago in https://github.com/ceph/ceph/pull/15632 (and long before that in https://github.com/ceph/ceph/pull/8241#issuecomment-220417387); i argued there that sharing ids made vstart more flexible, in that you could add/remove gateways after the initial `vstart.sh -n` without having to rewrite ceph.conf

jecluis commented:
@cbodley at the time, I needed this to properly test https://github.com/ceph/ceph/pull/25540 and running rgws with different ids was required for validation. I mean, I could make this conditional and not make it the default behavior of vstart.sh, but I can imagine this could be useful somewhere down the line.

cbodley commented:
@jecluis okay, makes sense. i think your approach can accomplish both goals, actually.. since vstart passes both --rgw-frontends and --admin-socket to radosgw on the command line, we don't actually rely on those [client.rgw.$id] sections in ceph.conf - so you could still create a cluster with RGW=1 and add more later

jecluis commented:
@cbodley forgot about this a while ago. I rebased, and changed the ids for the multiple rgws from made up hosts to their ports (think this is more straightforward). I believe it still performs pretty much the same way, but just a quick review would be nice.

cbodley commented:
looks great :+1:

for rgw multisite, we have a mstart.sh script that runs vstart multiple times in subdirectories, and a mrgw.sh that starts a gateway in a given cluster. mrgw.sh is still relying on the original client.rgw key, could you please fix that too? i got it working with this diff:
```diff
diff --git a/src/mrgw.sh b/src/mrgw.sh
index 7579f663be..cfb1ca0c36 100755
--- a/src/mrgw.sh
+++ b/src/mrgw.sh
@@ -28,4 +28,5 @@ logfile=$run_root/out/radosgw.${port}.log
 
 $vstart_path/mstop.sh $name radosgw $port
 
-$vstart_path/mrun $name radosgw --rgw-frontends="$rgw_frontend port=$port" -n client.rgw --pid-file=$pidfile --admin-socket=$asokfile "$@" --log-file=$logfile
+$vstart_path/mrun $name ceph -c $run_root/ceph.conf -k $run_root/keyring auth get-or-create client.rgw.$port mon 'allow rw' osd 'allow rwx' mgr 'allow rw' >> $run_root/keyring
+$vstart_path/mrun $name radosgw --rgw-frontends="$rgw_frontend port=$port" -n client.rgw.$port --pid-file=$pidfile --admin-socket=$asokfile "$@" --log-file=$logfile
```

jecluis commented:
@cbodley thanks. amended commit pushed.
